### ISSUE TITLE
Use of Attribute Routing possible?

### ISSUE THREAD

**Issue by jamesharling** [jbogard/ContosoUniversity#18]
I'm adapting this pattern in a test app I'm developing and I'm wondering if it's possible to use MVC's attribute routing? I'm not having much luck, I just get a 404 if I remove the convention-based routing. Any ideas?

**Comment by jbogard** [comment]
No idea, but it might be something with the controller factory?

**Comment by jamesharling** [comment]
I'm currently explicitly defining the custom controller factoy, in Global.asax:

`ControllerBuilder.Current.SetControllerFactory(typeof(ControllerFactory))`

The contents of the factoy are the same as your ContosoUniversity, bar a namespace change:

```
public class ControllerFactory : DefaultControllerFactory
{
    protected override Type GetControllerType(RequestContext requestContext, string controllerName)
    {
        return typeof(ControllerFactory).Assembly.GetType($"TestApp.Features.{controllerName}.UiController", throwOnError: false, ignoreCase: true);
    }
}
```

The `controllerName` argument passed to the factory is just `"Ui"`.

**Comment by jamesharling** [comment]
After doing some investigation, I've got attribute routing working by doing the following:
- Removing the `ControllerFactory` and its registration
- Removing the convention-based routing declarations

One caveat is that naming your controllers the same, e.g. `UiController`, won't work, as MVC tries to find views like `/Features/Ui/{ViewName}.cshtml`; I've named my controllers as `{FeatureName}Controller.cs` and this works, as the feature folder can be inferred from the controller name. If anyone knows of workaround to this, like a controller name attribute of sorts, I'd be interested!
